File: 1653-number-of-good-leaf-nodes-pairs/number-of-good-leaf-nodes-pairs.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// shortest path without custom weights = vanilla bfs should work

func countPairs(root *TreeNode, distance int) int {
    adjList := map[*TreeNode]map[*TreeNode]struct{}{}
    leafs := map[*TreeNode]struct{}{}
    var dfs func(r *TreeNode)
    dfs = func(r *TreeNode) {
        // base
        if r == nil {return}

        // logic
        if adjList[r] == nil {adjList[r] = map[*TreeNode]struct{}{}}
        if r.Left != nil {
            if adjList[r.Left] == nil {adjList[r.Left] = map[*TreeNode]struct{}{}}
            adjList[r][r.Left] = struct{}{}
            adjList[r.Left][r] = struct{}{}
            dfs(r.Left)

        }
        if r.Right != nil {
            if adjList[r.Right] == nil {adjList[r.Right] = map[*TreeNode]struct{}{}}
            adjList[r][r.Right] = struct{}{}
            adjList[r.Right][r] = struct{}{}        
            dfs(r.Right)

        }
        if r.Left == nil && r.Right == nil {
            leafs[r] = struct{}{}
            return
        }
    }
    dfs(root)
    count := 0
    for leaf, _ := range leafs {
        q := []*TreeNode{leaf}
        level := 0
        visited := map[*TreeNode]struct{}{}
        for len(q) != 0 {
            qSize := len(q)
            for qSize != 0 {
                dq := q[0]
                q = q[1:]
                if _, ok := leafs[dq]; ok && level <= distance && dq != leaf {
                    count++
                }
                for nei, _ := range adjList[dq] {
                    if _, ok := visited[nei]; !ok {
                        visited[nei] = struct{}{}
                        q = append(q, nei)
                    }
                }
                qSize--
            }
            level++
            if level > distance {break}
        }
    }
    return count / 2
}

File: 1653-number-of-good-leaf-nodes-pairs/number-of-good-leaf-nodes-pairs.py
class Solution:
    def _traverse_tree(self, curr_node, prev_node, graph, leaf_nodes):
        if curr_node is None:
            return
        if curr_node.left is None and curr_node.right is None:
            leaf_nodes.add(curr_node)
        if prev_node is not None:
            if prev_node not in graph:
                graph[prev_node] = []
            graph[prev_node].append(curr_node)

            if curr_node not in graph:
                graph[curr_node] = []
            graph[curr_node].append(prev_node)

        self._traverse_tree(curr_node.left, curr_node, graph, leaf_nodes)
        self._traverse_tree(curr_node.right, curr_node, graph, leaf_nodes)

    def countPairs(self, root, distance):
        graph = {}
        leaf_nodes = set()

        self._traverse_tree(root, None, graph, leaf_nodes)

        ans = 0

        for leaf in leaf_nodes:
            bfs_queue = []
            seen = set()
            bfs_queue.append(leaf)
            seen.add(leaf)
            for i in range(distance + 1):
                # Clear all nodes in the queue (distance i away from leaf node)
                # Add the nodes' neighbors (distance i+1 away from leaf node)
                size = len(bfs_queue)
                for j in range(size):
                    curr_node = bfs_queue.pop(0)
                    if curr_node in leaf_nodes and curr_node != leaf:
                        ans += 1
                    if curr_node in graph:
                        for neighbor in graph.get(curr_node):
                            if neighbor not in seen:
                                bfs_queue.append(neighbor)
                                seen.add(neighbor)
        return ans // 2